### ISSUE TITLE
Unexpected error when returning a filtered result with generic type guard

### ISSUE THREAD

**Issue by CodySchaaf** [Microsoft/TypeScript#30725]
I would have assumed this was just a limitation, and after searching open issues it didn't seem to quite match those which were related to just generic/type guard limitations. 

Since this works just by assigning it to a variable before returning, and also works if you are just using it in code and not trying to return it. 

**Summary**
When trying to construct a generic type guard for using with discriminated unions, the type is correctly inferred, but if assigned to a variable--or returned in a function--with that explicit type annotation I get an error. 

**Search Terms:**
discriminated union return type unexpected behavior with generic type guard
discriminated union 
generic type guard

**Code**

```ts
export const isOfType = <TValue extends string| symbol, T extends ({ type: any })>(type: TValue) =>
	(objectToNarrow: T|undefined): objectToNarrow is Extract<NonNullable<T>, { type: TValue }> =>
		!!objectToNarrow && objectToNarrow.type === type;

const CatType = Symbol('cat');
const DogType = Symbol('dog');

interface Cat {
	readonly type: typeof CatType;
	purr: () => void;
}

interface Dog {
	readonly type: typeof DogType;
}

type Animal = Cat | Dog;

const myCat: Array<Animal | undefined> = [{type: CatType, purr: () => { console.log('purr'); }}];


/**
 * Argument of type '(objectToNarrow: Cat | undefined) => objectToNarrow is Cat' is not assignable to parameter of type '(value: Cat | Dog | undefined, index: number, array: (Cat | Dog | undefined)[]) => any'.
 * Types of parameters 'objectToNarrow' and 'value' are incompatible.
 * Type 'Cat | Dog | undefined' is not assignable to type 'Cat | undefined'.
 * Property 'purr' is missing in type 'Dog' but required in type 'Cat'.
 */
const getCat1 = (): Cat[] => {
	return myCat.filter(isOfType(CatType)) // error here
};

const getCat2 = (): Cat[] => {
	const foo = myCat.filter(isOfType(CatType));
	return foo; // but no error here
};

const cat1 = getCat1();
const cat2 = getCat2();
```

**Expected behavior:**
Either both functions should work, or neither should work. 

**Actual behavior:**
Only one of the functions work, but it just assigns the value to a variable before returning.

**Playground Link:** 
Note: enable `strictFunctionTypes`

https://www.typescriptlang.org/play/#src=export%20const%20isOfType%20%3D%20%3CTValue%20extends%20string%7C%20symbol%2C%20T%20extends%20(%7B%20type%3A%20any%20%7D)%3E(type%3A%20TValue)%20%3D%3E%0A%09(objectToNarrow%3A%20T%7Cundefined)%3A%20objectToNarrow%20is%20Extract%3CNonNullable%3CT%3E%2C%20%7B%20type%3A%20TValue%20%7D%3E%20%3D%3E%0A%09%09!!objectToNarrow%20%26%26%20objectToNarrow.type%20%3D%3D%3D%20type%3B%0A%0Aconst%20CatType%20%3D%20Symbol('cat')%3B%0Aconst%20DogType%20%3D%20Symbol('dog')%3B%0A%0Ainterface%20Cat%20%7B%0A%09readonly%20type%3A%20typeof%20CatType%3B%0A%09purr%3A%20()%20%3D%3E%20void%3B%0A%7D%0A%0Ainterface%20Dog%20%7B%0A%09readonly%20type%3A%20typeof%20DogType%3B%0A%7D%0A%0Atype%20Animal%20%3D%20Cat%20%7C%20Dog%3B%0A%0Aconst%20myCat%3A%20Array%3CAnimal%20%7C%20undefined%3E%20%3D%20%5B%7Btype%3A%20CatType%2C%20purr%3A%20()%20%3D%3E%20%7B%20console.log('purr')%3B%20%7D%7D%5D%3B%0A%0A%0A%2F**%0A%20*%20Argument%20of%20type%20'(objectToNarrow%3A%20Cat%20%7C%20undefined)%20%3D%3E%20objectToNarrow%20is%20Cat'%20is%20not%20assignable%20to%20parameter%20of%20type%20'(value%3A%20Cat%20%7C%20Dog%20%7C%20undefined%2C%20index%3A%20number%2C%20array%3A%20(Cat%20%7C%20Dog%20%7C%20undefined)%5B%5D)%20%3D%3E%20any'.%0A%20*%20Types%20of%20parameters%20'objectToNarrow'%20and%20'value'%20are%20incompatible.%0A%20*%20Type%20'Cat%20%7C%20Dog%20%7C%20undefined'%20is%20not%20assignable%20to%20type%20'Cat%20%7C%20undefined'.%0A%20*%20Property%20'purr'%20is%20missing%20in%20type%20'Dog'%20but%20required%20in%20type%20'Cat'.%0A%20*%2F%0Aconst%20getCat1%20%3D%20()%3A%20Cat%5B%5D%20%3D%3E%20%7B%0A%09return%20myCat.filter(isOfType(CatType))%20%2F%2F%20error%20here%0A%7D%3B%0A%0Aconst%20getCat2%20%3D%20()%3A%20Cat%5B%5D%20%3D%3E%20%7B%0A%09const%20foo%20%3D%20myCat.filter(isOfType(CatType))%3B%0A%09return%20foo%3B%20%2F%2F%20but%20no%20error%20here%0A%7D%3B%0A%0Aconst%20cat1%20%3D%20getCat1()%3B%0Aconst%20cat2%20%3D%20getCat2()%3B%0A

**Repository with Repo**
https://github.com/CodySchaaf/typescript-issue

**Comment by CodySchaaf** [comment]
Because this is related to `strictFunctionTypes`, I'm going to assume I just don't understand the intricacies of contravariantly vs bivariantly, but I would still expect both to error under this or neither. And there is probably a related issue I just couldn't find. Sorry in advance for the inevitable "This is a dupe comment"

**Comment by jack-williams** [comment]
Writing `isOfType` like this should fix it:
```ts
export const isOfType = <TValue extends string| symbol>(type: TValue) =>
	<T extends ({ type: any })>(objectToNarrow: T|undefined): objectToNarrow is Extract<NonNullable<T>, { type: TValue }> =>
		!!objectToNarrow && objectToNarrow.type === type;
```

**Comment by CodySchaaf** [comment]
Adding `T & {}` seems to have done the trick thanks! The first suggestion caused issues when trying to use this down stream in a custom rxjs operator where it was unable to reconcile `objectToNarrow` and `{}`'s missing value parameter. 

```ts

export const partitionByType = <TValue1 extends string | symbol, TValue2 extends string | symbol, T extends ({ type: any })>(
  type1: TValue1,
  type2: TValue2,
) => (source: Observable<T|undefined>): Observable<[
  ExtractDefinedType$<T, TValue1>,
  ExtractDefinedType$<T, TValue2>
  ]> => {
  const sharedSource = source.pipe(share());
  return of<[ExtractDefinedType$<T, TValue1>, ExtractDefinedType$<T, TValue2>]>([
    filter(isOfType<TValue1, T>(type1))(sharedSource),
    filter(isOfType<TValue2, T>(type2))(sharedSource),
  ]);
};
```

So I'll stick with `T & {}` Thanks a lot for the simple solution.
